### DISCUSSION TITLE
Sublattice site ratios for simple binary fluid system

### DISCUSSION THREAD

**Discussion by chuabinghong** [PhasesResearchLab/ESPEI#242] [Q&A] [answered]
Hey all,

I'm trying to prepare a .tdb for a simple A-B binary fluid system with specific heat data at varying temperature and compositions. When trying run_espei(), the following error is raised:

`espei.datasets.DatasetError: Dataset Error in input_data\184.90K.json: Sublattice configuration ['A', 'B'] and sublattice site ratio [1] describe different numbers of sublattices (2 and 1).`

what should the correct sublattice ratios be in the input.json and data.json files? I defaulted to [1] as per the examples found in https://espei.org/en/latest/input_data.html

Here's the input.json (A is NH3):
```json
{  
"components": ["A", "B"],
  "phases": {
    "LIQUID": {
      "sublattice_model": [[ "A","B"]],
      "sublattice_site_ratios": [1]
    }
  }
}
```

And an example of the data.json:
```json
{
  "reference": "test",
  "components": ["A", "B"],
  "phases": ["LIQUID"],
  "solver": {
    "mode": "manual",
    "sublattice_site_ratios": [1],
    "sublattice_configurations": [["A", "B"]],
    "sublattice_occupancies": [[[0.311602715175133, 0.688397284824867]]],
    "comment": "-"
  },
  "conditions": {
    "P": 101325,
    "T": 184.896586
  },
  "output": "CPM",
  "values": [[[39.77684022357711]]]
}
```

Attached is a reproducible example.
[code.zip](https://github.com/PhasesResearchLab/ESPEI/files/9383598/code.zip)

**Comment by bocklund** [comment, marked as answer]
#### Short answer 
I believe the cause is that sublattice configurations should be a list of configurations. The simple fix is to change 

```json
    "sublattice_configurations": [["A", "B"]],
```

to 

```
    "sublattice_configurations": [[["A", "B"]]],
```

The important thing to keep in mind is that the `"sublattice_configurations"` and `"sublattice_occupancies"` should have exactly the same shape, with configurations having the species names and occupancies having their site fractions. 


#### Longer answer

`"sublattice_configurations"` is a list of configurations, so 

```json
    "sublattice_configurations": [
        [["A", "B"]],  # configuration 1
        [["A", "B"]],  # configuration 2
        #  ...
    ],
```

and there's one historical piece of notation that you can specify a singly-occupied sublattice without using an extra pair of brackets. That is:

```json
    "sublattice_configurations": [
        [["A"]],  # configuration 1, pure A
        [["B"]],  # configuration 2, pure B
        [["A", "B"]],  # configuration 3, mised
        #  ...
    ],
```

is equivalent to:

```json
    "sublattice_configurations": [
        ["A"],  # configuration 1, pure A
        ["B"],  # configuration 2, pure B
        [["A", "B"]],  # configuration 3, mised
        #  ...
    ],
```

and so writing 

```json
    "sublattice_configurations": [["A", "B"]],
```

actually gets parsed as a phase with two sublattices, causing this issue. Eventually I want to remove this behavior, but it'll be a breaking change so it will stay for the foreseeable future.

**Reply by chuabinghong** [reply]
That solved the issue, thanks!

However, a new error is raised further down the line:
```
File ".\espei\datasets.py", line 74, in check_dataset
    is_equilibrium = 'solver' not in dataset.keys() and dataset['output'] != 'ZPF'
KeyError: 'output'
```

How should I proceed from here? 
(Do let me know if I should open a new discussion thread for this)

**Reply by bocklund** [reply]
It could be a new thread, but I can answer it here 😄 

From your zipped reproducible example, you have the JSON datasets organized alongside the phase models file. This error is caused by ESPEI trying to read your phase models file like a dataset, when it is not. The fix is to organize your datasets in their own directory. 

The directory structure of your reproducible example looks like this:

```
.
├── input_data
│   ├── 184.90K.json
│   ├── 184.95K.json
│   ├── espei-in.yaml
│   └── input.json
└── test_b.py
```

instead it should look like:

```
.
├── espei-in.yaml
├── input.json
├── input_data
│   ├── 184.90K.json
│   └── 184.95K.json
└── test_b.py
```

if you run ESPEI from this directory, `espei --in espei-in.yaml` it will work. Or to modify your `test_b.py` to be:

```python
# test_b.py
from espei import run_espei
import yaml

##
with open(r'espei-in.yaml') as fp:
    inputyaml = yaml.safe_load(fp)

dbf = run_espei(inputyaml)
```

then `python test_b.py`.

Running that, I get the output:

```
WARNING:espei.database_utils - The reference phase for B, BETA_RHOMBO_B, is not in the supplied phase models and won't be added to the Database phases. Fitting formation energies will not be possible.
WARNING:espei.database_utils - A has no entry in the SGTE91SER reference data. Fitting formation energies will not be possible.
Traceback (most recent call last):
  File "/Users/bocklund1/Downloads/code/test_b.py", line 9, in <module>
    dbf = run_espei(inputyaml)
  File "/Users/bocklund1/src/calphad/espei/espei/espei_script.py", line 177, in run_espei
    dbf = generate_parameters(phase_models, datasets, refdata, excess_model,
  File "/Users/bocklund1/src/calphad/espei/espei/paramselect.py", line 518, in generate_parameters
    dbf = initialize_database(phase_models, ref_state, dbf)
  File "/Users/bocklund1/src/calphad/espei/espei/database_utils.py", line 78, in initialize_database
    el_ser_data["phase"] = aliases.get(el_ser_data["phase"], el_ser_data["phase"])
KeyError: 'phase'
```

Which is expected because ESPEI doesn't know about elements `A` or `B` in the SGTE91 reference data. If A and B will be replaced with real, pure elements that are in the SGTE91 reference data, you can replace them and proceed. Otherwise, you'll want to take a look at the following link to see how to provide your own unary data to ESPEI https://espei.org/en/latest/howto_use_custom_unary.html, or you can provide your own TDB with elements A and B defined already, which may be easier if you have some experience with TDB files already. Please open another thread for that one if you're having trouble.

**Reply by chuabinghong** [reply]
Thank you for the clarifications.

I'm in the midst of creating a custom unary data, and will write a new thread if/when I face challenges there.
